Failed to compile the project with `kotlinx.coroutines` dependency

Libraries published with: Konan: 0.8-dev-2598 / Kotlin: 1.2.60

Reproducer: https://github.com/Kotlin/kotlinx-io/tree/native
Run: `./gradlew clean assemble`

```
 * Compiler version info: Konan: 0.8-dev-2598 / Kotlin: 1.2.60
 * Output kind: LIBRARY

exception: java.lang.AssertionError: internal final fun onStartInternal(): kotlin.Unit defined in kotlinx.coroutines.experimental.AbstractCoroutine[DeserializedSimpleFunctionDescriptor@6010d0d1]
        at org.jetbrains.kotlin.backend.konan.llvm.BinaryInterfaceKt.getSymbolName(BinaryInterface.kt:204)
        at org.jetbrains.kotlin.backend.konan.llvm.ContextUtils$DefaultImpls.getLlvmFunction(ContextUtils.kt:171)
        at org.jetbrains.kotlin.backend.konan.llvm.RTTIGenerator.getLlvmFunction(RTTIGenerator.kt:29)
        at org.jetbrains.kotlin.backend.konan.llvm.ContextUtils$DefaultImpls.getEntryPointAddress(ContextUtils.kt:183)
        at org.jetbrains.kotlin.backend.konan.llvm.RTTIGenerator.getEntryPointAddress(RTTIGenerator.kt:29)
        at org.jetbrains.kotlin.backend.konan.llvm.RTTIGenerator.methodTableRecords(RTTIGenerator.kt:257)
        at org.jetbrains.kotlin.backend.konan.llvm.RTTIGenerator.generate(RTTIGenerator.kt:196)
        at org.jetbrains.kotlin.backend.konan.llvm.RTTIGeneratorVisitor.visitClass(IrToBitcode.kt:165)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitClass(IrElementVisitorVoid.kt:44)
        at org.jetbrains.kotlin.backend.konan.llvm.RTTIGeneratorVisitor.visitClass(IrToBitcode.kt:147)
        at org.jetbrains.kotlin.backend.konan.llvm.RTTIGeneratorVisitor.visitClass(IrToBitcode.kt:147)
        at org.jetbrains.kotlin.ir.declarations.impl.IrClassImpl.accept(IrClassImpl.kt:100)
        at org.jetbrains.kotlin.ir.declarations.impl.IrFileImpl.acceptChildren(IrFileImpl.kt:79)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptChildrenVoid(IrElementVisitorVoid.kt:245)
        at org.jetbrains.kotlin.backend.konan.llvm.RTTIGeneratorVisitor.visitElement(IrToBitcode.kt:153)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitPackageFragment(IrElementVisitorVoid.kt:30)
        at org.jetbrains.kotlin.backend.konan.llvm.RTTIGeneratorVisitor.visitPackageFragment(IrToBitcode.kt:147)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitFile(IrElementVisitorVoid.kt:37)
        at org.jetbrains.kotlin.backend.konan.llvm.RTTIGeneratorVisitor.visitFile(IrToBitcode.kt:147)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitFile(IrElementVisitorVoid.kt:38)
        at org.jetbrains.kotlin.backend.konan.llvm.RTTIGeneratorVisitor.visitFile(IrToBitcode.kt:147)
        at org.jetbrains.kotlin.backend.konan.llvm.RTTIGeneratorVisitor.visitFile(IrToBitcode.kt:147)
        at org.jetbrains.kotlin.ir.declarations.impl.IrFileImpl.accept(IrFileImpl.kt:76)
        at org.jetbrains.kotlin.ir.declarations.impl.IrModuleFragmentImpl.acceptChildren(IrModuleFragmentImpl.kt:49)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptChildrenVoid(IrElementVisitorVoid.kt:245)
        at org.jetbrains.kotlin.backend.konan.llvm.RTTIGeneratorVisitor.visitElement(IrToBitcode.kt:153)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitModuleFragment(IrElementVisitorVoid.kt:27)
        at org.jetbrains.kotlin.backend.konan.llvm.RTTIGeneratorVisitor.visitModuleFragment(IrToBitcode.kt:147)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitModuleFragment(IrElementVisitorVoid.kt:28)
        at org.jetbrains.kotlin.backend.konan.llvm.RTTIGeneratorVisitor.visitModuleFragment(IrToBitcode.kt:147)
        at org.jetbrains.kotlin.backend.konan.llvm.RTTIGeneratorVisitor.visitModuleFragment(IrToBitcode.kt:147)
        at org.jetbrains.kotlin.ir.declarations.impl.IrModuleFragmentImpl.accept(IrModuleFragmentImpl.kt:46)
        at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptVoid(IrElementVisitorVoid.kt:241)
        at org.jetbrains.kotlin.backend.konan.llvm.IrToBitcodeKt$emitLLVM$1.invoke(IrToBitcode.kt:65)
        at org.jetbrains.kotlin.backend.konan.llvm.IrToBitcodeKt$emitLLVM$1.invoke(IrToBitcode.kt)
        at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:139)
        at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:118)
        at org.jetbrains.kotlin.konan.util.UtilKt.profileIf(Util.kt:34)
        at org.jetbrains.kotlin.backend.konan.PhaseManager.phase$backend_native_compiler(KonanPhases.kt:138)
        at org.jetbrains.kotlin.backend.konan.llvm.IrToBitcodeKt.emitLLVM(IrToBitcode.kt:64)
        at org.jetbrains.kotlin.backend.konan.KonanDriverKt$runTopLevelPhases$4$2.invoke(KonanDriver.kt:97)
        at org.jetbrains.kotlin.backend.konan.KonanDriverKt$runTopLevelPhases$4$2.invoke(KonanDriver.kt)
        at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:139)
        at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:118)
        at org.jetbrains.kotlin.konan.util.UtilKt.profileIf(Util.kt:34)
        at org.jetbrains.kotlin.backend.konan.PhaseManager.phase$backend_native_compiler(KonanPhases.kt:138)
        at org.jetbrains.kotlin.backend.konan.KonanDriverKt$runTopLevelPhases$4.invoke(KonanDriver.kt:96)
        at org.jetbrains.kotlin.backend.konan.KonanDriverKt$runTopLevelPhases$4.invoke(KonanDriver.kt)
        at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:139)
        at org.jetbrains.kotlin.backend.konan.PhaseManager$phase$$inlined$with$lambda$1.invoke(KonanPhases.kt:118)
        at org.jetbrains.kotlin.konan.util.UtilKt.profileIf(Util.kt:34)
        at org.jetbrains.kotlin.backend.konan.PhaseManager.phase$backend_native_compiler(KonanPhases.kt:138)
        at org.jetbrains.kotlin.backend.konan.KonanDriverKt.runTopLevelPhases(KonanDriver.kt:90)
        at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:75)
        at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:42)
        at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.java:95)
        at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.java:50)
        at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:88)
        at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:66)
        at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:34)
        at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMainNoExit(CLITool.kt:180)
        at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMain(CLITool.kt:171)
        at org.jetbrains.kotlin.cli.bc.K2Native$Companion$main$1.invoke(K2Native.kt:197)
        at org.jetbrains.kotlin.cli.bc.K2Native$Companion$main$1.invoke(K2Native.kt:188)
        at org.jetbrains.kotlin.konan.util.UtilKt.profileIf(Util.kt:34)
        at org.jetbrains.kotlin.konan.util.UtilKt.profile(Util.kt:29)
        at org.jetbrains.kotlin.cli.bc.K2Native$Companion.main(K2Native.kt:190)
        at org.jetbrains.kotlin.cli.bc.K2NativeKt.main(K2Native.kt:202)
        at org.jetbrains.kotlin.cli.utilities.MainKt.main(main.kt:27)
```

--- a/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/BinaryInterface.kt
+++ b/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/BinaryInterface.kt
@@ -24,11 +24,14 @@ import org.jetbrains.kotlin.backend.konan.library.KonanLibraryReader
 import org.jetbrains.kotlin.backend.konan.optimizations.DataFlowIR
 import org.jetbrains.kotlin.descriptors.ModuleDescriptor
 import org.jetbrains.kotlin.descriptors.PropertyAccessorDescriptor
+import org.jetbrains.kotlin.descriptors.Visibilities
+import org.jetbrains.kotlin.descriptors.Visibility
 import org.jetbrains.kotlin.descriptors.annotations.AnnotationDescriptor
 import org.jetbrains.kotlin.ir.declarations.*
 import org.jetbrains.kotlin.ir.symbols.IrTypeParameterSymbol
 import org.jetbrains.kotlin.ir.types.*
 import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
 import org.jetbrains.kotlin.resolve.constants.StringValue
 import org.jetbrains.kotlin.utils.addToStdlib.ifNotEmpty
 
@@ -96,7 +99,7 @@ internal tailrec fun DeclarationDescriptor.isExported(): Boolean {
         else -> null
     }
 
-    if (visibility != null && !visibility.isPublicAPI) {
+    if (visibility != null && !visibility.isPublicAPI && visibility != Visibilities.INTERNAL) {
         // If the declaration is explicitly marked as non-public,
         // then it must not be accessible from other modules.
         return false
@@ -194,9 +197,21 @@ internal val FunctionDescriptor.functionName: String
                 }
             }
 
+            val name = this.name.mangleIfInternal(this.module, this.visibility)
+
             return "$name$signature"
         }
     }
+
+private fun Name.mangleIfInternal(moduleDescriptor: ModuleDescriptor, visibility: Visibility): String =
+        if (visibility != Visibilities.INTERNAL) {
+            this.asString()
+        } else {
+            val moduleName = moduleDescriptor.name.asString()
+                    .let { it.substring(1, it.lastIndex) } // Remove < and >.
+
+            "$this\$$moduleName"
+        }
 
 internal val FunctionDescriptor.symbolName: String
     get() {

